### PR TITLE
raft: add DisableProposalForwarding option

### PR DESCRIPTION
**Pull request by smetro**
this allows users to disable followers from forwarding proposals to the
leader.

# Contributing guidelines

Please read our [contribution workflow][contributing] before submitting a pull request.

[contributing]: https://github.com/coreos/etcd/blob/master/CONTRIBUTING.md#contribution-flow

### CODE REVIEW

**Comment by xiang90** [comment]
@smetro Can you explain your use case and motivation for doing this (I can guess I why, but we need to keep the record here). Also we need to document the option better. So users know when they want to enable this feature.

**Comment by smetro** [comment]
The use case is that when a node in my Raft cluster receives a client request I add an [HLC timestamp](http://muratbuffalo.blogspot.ca/2014/07/hybrid-logical-clocks.html) to the request before passing the data to [node.Propose()](https://github.com/coreos/etcd/blob/master/raft/node.go#L406). That timestamp needs to be computed by the Raft leader so that it remains monotonically increasing (I maintain monotonicity after a failover by advertising these HLC timestamps from the leader to the followers, so that a new leader starts from a valid HLC timestamp). If a follower adds the timestamp and forwards it to the leader, the leader can't update the timestamp before creating the AppendEntry request.

**Comment by smetro** [comment]
Will close/reopen to try to retrigger the 'continuous-integration/travis-ci/pr', since I've removed whitespace from the commit message.

**Comment by smetro** [comment]
Will close/reopen to try to retrigger the 'jenkins-cov', since https://github.com/coreos/etcd/issues/7980 is closed.

File: raft/raft.go [hunk 1]
```diff
@@ -176,6 +176,10 @@ type Config struct {
 	// Logger is the logger used for raft log. For multinode which can host
 	// multiple raft group, each raft group can have its own logger
 	Logger Logger
+
+	// DisableProposalForwarding set to true means that followers will drop
+	// proposals, rather than forwarding them to the leader.
+	DisableProposalForwarding bool
```
Comment (xiang90) [inline raft/raft.go]: can you provide an example use case here to justify this feature flag?

**Comment by xiang90** [comment]
I am fine with merging this after providing an example use case in the comment. thanks!

**Comment by smetro** [comment]
Thanks, I added a comment trying to explain the use case, please let me know what you think.

**Comment by xiang90** [comment]
LGTM. Can you squash the two commits into one? And follow the commit message style? (raft: xxx)

So CI can pass, and I can get your pr merged.

**Comment by smetro** [comment]
Thanks, I have squashed them.

**Pull request merged**